src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,176 @@
+import functools
+import logging
+from typing import Any, Dict, MutableMapping, Optional, Sequence
+
+from arroyo import Topic
+from arroyo.backends.kafka import KafkaConsumer
+from arroyo.processing import StreamProcessor
+from arroyo.processing.strategies import ProcessingStrategy as ProcessingStep
+from arroyo.processing.strategies.streaming import KafkaConsumerStrategyFactory
+from confluent_kafka import Producer
+from django.conf import settings
+
+from sentry.utils import json, kafka_config
+
+DEFAULT_QUEUED_MAX_MESSAGE_KBYTES = 50000
+DEFAULT_QUEUED_MIN_MESSAGES = 10000
+
+logger = logging.getLogger(__name__)
+
+
+def initializer():
+    from sentry.runner import configure
+
+    configure()
+
+
+def get_config(topic: str, **options) -> MutableMapping[str, Any]:
+    consumer_config = kafka_config.get_kafka_consumer_cluster_options(
+        "default",
+        override_params={
+            "enable.auto.commit": False,
+            "enable.auto.offset.store": False,
+            "group.id": "ingest-metrics-consumer",
+            "default.topic.config": {"auto.offset.reset": "latest"},
+            # overridden to reduce memory usage when there's a large backlog
+            "queued.max.messages.kbytes": DEFAULT_QUEUED_MAX_MESSAGE_KBYTES,
+            "queued.min.messages": DEFAULT_QUEUED_MIN_MESSAGES,
+        },
+    )
+    return consumer_config
+
+
+class FlushBatchStep(ProcessingStep):
+    def __init__(self) -> None:
+        snuba_metrics = settings.KAFKA_TOPICS[settings.KAFKA_SNUBA_METRICS]
+        snuba_metrics_producer = Producer(
+            kafka_config.get_kafka_producer_cluster_options(snuba_metrics["cluster"]),
+        )
+        self.__producer = snuba_metrics_producer
+        self.__producer_topic = settings.KAFKA_TOPICS[settings.KAFKA_SNUBA_METRICS].get(
+            "topic", "snuba-metrics"
+        )
+
+        self.__messages = []
+        self.__closed = False
+
+    def poll(self) -> None:
+        pass
+
+    def submit(self, message: Any) -> None:
+        assert not self.__closed
+
+        self.__producer.produce(
+            topic=self.__producer_topic,
+            key=None,
+            value=json.dumps(message.payload).encode(),
+            on_delivery=self._callback,
+        )
+
+        messages_left = self.__producer.flush(1.0)
```

Comment:
See the comment below for detail. 
As per arroyo docs, the `submit` method should not do waiting operations.
It makes sense to run the asynchronous operation here (`produce`) and check the callbacks in the `poll` method.

Specifically `poll` would not wait. it would only count the number of callbacks that have returned positively (they are futures, so there is a way to check if the callback returned without waiting), and once the batch size is reached, commit in Kafka. 
It also allows you to commit up to the first message that failed.